ambari-server/src/main/java/org/apache/ambari/server/events/MessageEmitter.java

```diff
@@ -17,70 +17,248 @@
+    event.setMessageId(MESSAGE_ID.getAndIncrement());
+
+    Long hostId = event.getHostId();
+    if (!messagesToEmit.containsKey(hostId)) {
+      LOG.error("Trying to emit message to unregistered host with id {}", hostId);
+      return;
+    }
+    messagesToEmit.get(hostId).add(event);
+  }
+
+  private class MessagesToEmitMonitor implements Runnable {
+
+    private final Long hostId;
+
+    public MessagesToEmitMonitor(Long hostId) {
+      this.hostId = hostId;
+    }
+
+    @Override
+    public void run() {
+      while (true) {
+        try {
+          ExecutionCommandEvent event = messagesToEmit.get(hostId).take();
+          EmitMessageTask emitMessageTask = new EmitMessageTask(event);
+          ScheduledFuture scheduledFuture =
+              emittersExecutor.scheduleAtFixedRate(emitMessageTask,
+                  0, retryInterval, TimeUnit.SECONDS);
+          emitMessageTask.setScheduledFuture(scheduledFuture);
+          unconfirmedMessages.put(event.getMessageId(), scheduledFuture);
+
+          scheduledFuture.get();
+        } catch (InterruptedException e) {
+          // can be interrupted when no responses were received from agent and HEARTBEAT_LOST will be fired
+          return;
+        } catch (CancellationException e) {
+          // scheduled tasks can be canceled
+          e.printStackTrace();
+        } catch (ExecutionException e) {
+          e.printStackTrace();
+          // generate delivery failed event
+          ambariEventPublisher.publish(new HeartbeatLostEvent(hostId));
+          return;
+        }
+      }
+    }
+  }
+
+  public void processReceiveReport(AckReport ackReport) {
+    Long messageId = ackReport.getMessageId();
+    if (AckReport.AckStatus.OK.equals(ackReport.getStatus())) {
+      if (unconfirmedMessages.containsKey(messageId)) {
+        unconfirmedMessages.get(messageId).cancel(true);
+        unconfirmedMessages.remove(messageId);
+      } else {
+        LOG.warn("OK agent report was received again for already complete command with message id {}", messageId);
+      }
+    } else {
+      LOG.error("Received {} agent report for execution command with messageId {} with following reason: {}",
+          ackReport.getStatus(), messageId, ackReport.getReason());
+    }
+  }
+
+  private class EmitMessageTask implements Runnable {
+
+    private final ExecutionCommandEvent executionCommandEvent;
+    private ScheduledFuture scheduledFuture;
+    private int retry_counter = 0;
+
+    public EmitMessageTask(ExecutionCommandEvent executionCommandEvent) {
+      this.executionCommandEvent = executionCommandEvent;
+    }
+
+    public void setScheduledFuture(ScheduledFuture scheduledFuture) {
+      this.scheduledFuture = scheduledFuture;
+    }
+
+    @Override
+    public void run() {
+      if (retry_counter >= retryCount) {
+        // generate delivery failed event and cancel emitter
+        ambariEventPublisher.publish(new HeartbeatLostEvent(executionCommandEvent.getHostId()));
```

Comment:
Are we overloading the HB lost event here? Why don't we have a CommandFailedEvent here?